### PR TITLE
Refactor naming and serialization for PeerIDs

### PR DESCRIPTION
**Pull request by borzunov**
This PR follows #323 and does the remaining mass refactors (we've agreed on the most of them in comments to #323):

1. Rename `Endpoint` to `PeerID` in averager (+ related variable names)
2. Rename the `P2P.id` field to `P2P.peer_id` (because the local peer ID is stored in the `.peer_id` fields in all other classes)
3. Serialize `PeerID`s as `bytes` instead of Base58 string
4. Remove `JoinRequest.peer_id` and `AveragingData.peer_id` fields (they duplicate `context.remote_id`)
5. Remove the `DecentralizedAveraging` gRPC interface (not used anymore)

### CODE REVIEW

File: hivemind/averaging/allreduce.py [hunk 1]
```diff
@@ -74,60 +74,60 @@ def __init__(
...
 
-        endpoint_index = self.ordered_group_endpoints.index(self.endpoint)
+        peer_id_index = self.ordered_group_peer_ids.index(self.peer_id)
         self.tensor_part_container = TensorPartContainer(tensors, peer_fractions, **kwargs)
-        self.parts_for_local_averaging = self.tensor_part_container.get_raw_input_parts(endpoint_index)
+        self.parts_for_local_averaging = self.tensor_part_container.get_raw_input_parts(peer_id_index)
         self.tensor_part_reducer = TensorPartReducer(
             tuple(part.shape for part in self.parts_for_local_averaging),
-            len(self.sender_endpoints),
+            len(self.sender_peer_ids),
```
Comment (justheuristic) [inline hivemind/averaging/allreduce.py:99]: nit: sender_ids?

File: hivemind/averaging/allreduce.py [hunk 2]
```diff
@@ -38,11 +38,11 @@ class AllReduceRunner(ServicerBase):
     :param group_id: unique identifier of this specific all-reduce run
     :param tensors: local tensors that should be averaged with groupmates
     :param tensors: local tensors that should be averaged with groupmates
-    :param endpoint: your endpoint, must be included in ordered_group_endpoints
-    :param ordered_group_endpoints: group endpoints ordered s.t. i-th endpoint is responsible for averaging i-th part
+    :param peer_id: your peer_id, must be included in ordered_group_peer_ids
+    :param ordered_group_peer_ids: group peer_ids ordered s.t. i-th peer_id is responsible for averaging i-th part
```
Comment (justheuristic) [inline hivemind/averaging/allreduce.py:42]: nit: ordered_peer_ids ?

**Review by justheuristic — APPROVED** [review]

File: hivemind/averaging/matchmaking.py [hunk 3]
```diff
@@ -243,16 +243,16 @@ async def rpc_join_group(
         self, request: averaging_pb2.JoinRequest, _context: P2PContext
     ) -> AsyncIterator[averaging_pb2.MessageFromLeader]:
         """accept or reject a join request from another averager; if accepted, run him through allreduce steps"""
-        request_endpoint = None
+        request_peer_id = None
```
Comment (justheuristic) [inline hivemind/averaging/matchmaking.py:246]: nit: maybe just peer_id?

Comment (borzunov) [reply hivemind/averaging/matchmaking.py]: Changed to `remote_peer_id` (to emphasize that it is not the local peer ID)

File: hivemind/proto/averaging.proto [hunk 4]
```diff
@@ -32,7 +25,6 @@ enum MessageCode {
 }
 
 message JoinRequest {
-  string endpoint = 1;          // A follower accepts incoming allreduce requests at this address
   bytes schema_hash = 2;        // A hash that describes follower's tensors (shapes, num tensors, etc)
```
Comment (justheuristic) [inline hivemind/proto/averaging.proto:28]: [the resulting protobuf starts from index=2; discussed with @borzunov and decided to keep it that way unless someone has strong objections]

**Review by justheuristic — APPROVED** [review]

File: benchmarks/benchmark_averaging.py [hunk 5]
```diff
@@ -70,7 +70,7 @@ def run_averager(index):
         processes.update({dht, averager})
 
         logger.info(
-            f"Averager {index}: started on endpoint {averager.endpoint}, group_bits: {averager.get_group_bits()}"
+            f"Averager {index}: started on peer id {averager.peer_id}, group_bits: {averager.get_group_bits()}"
```
Comment (justheuristic) [inline benchmarks/benchmark_averaging.py:73]:
```suggestion
            f"Averager {index}: started with peer id {averager.peer_id}, group_bits: {averager.get_group_bits()}"
```

**Review by yhn112 — APPROVED** [review]

**Pull request merged**